### PR TITLE
Add Scanf.{s,b,}scanf_opt

### PR DESCRIPTION
**Pull request by nojb**
This PR adds option-returning variants to `Scanf.sscanf` and friends. This function can raise no fewer than 4 exceptions, three of which correspond to scanning failure. The new variants return `None` in this case. `Invalid_argument` (corresponding to an invalid format string) continues to be raised, as usual.

Tests will be added in due course, assuming the proposal is well-received. **EDIT:** Some tests have been added.

Not sure that the type signature (using the new `_ scanner_opt`) is the most efficient one: I was forced to add an extra type variable, can this be avoided? Suggestions welcome. **EDIT:** The signature has been fixed.

### CODE REVIEW

**Comment by Octachron** [comment]
The type issue reflects that your current implementation is not handling `%r` specifiers, for instance
```ocaml
 Scanf.sscanf_opt "1 2 4" "%d %r %d" 
  (fun b -> Scanf.bscanf b "%d" (fun r -> r))
  (fun x y z -> x + y + z)
```

**Comment by nojb** [comment]
> The type issue reflects that your current implementation is not handling `%r` specifiers, for instance

Ah, I knew something was fishy :) Thanks, I will look into it.

**Comment by nojb** [comment]
> The type issue reflects that your current implementation is not handling `%r` specifiers, for instance
> 
> ```ocaml
>  Scanf.sscanf_opt "1 2 4" "%d %r %d" 
>   (fun b -> Scanf.bscanf b "%d" (fun r -> r))
>   (fun x y z -> x + y + z)
> ```

I pushed another version that should behave better :) Thanks!

**Review by gasche — APPROVED** [review]
I believe that the addition makes sense, and that the implementation is correct.

**Comment by nojb** [comment]
Thanks @gasche for the review. As this touches the stdlib we need a second official review. @Octachron: would you be ablet to do that (as you have already taken a look at this PR)?

**Comment by gasche** [comment]
(Note: I did ask myself many questions about this API, like "should we ask for `('a, exn) result` instead of `'a option`? But I decided to do best-effort resolution of these questions *internally* instead of inflicting them onto the next reviewer.)

(Another question: "should the generic function return a `(<args>, error) result` instead of taking a success and a failure continuation?)

**Comment by Octachron** [comment]
Sure, I shall do a full review sometimes this week.

File: stdlib/scanf.mli [hunk 1]
```diff
@@ -247,6 +250,11 @@ val bscanf : Scanning.in_channel -> ('a, 'b, 'c, 'd) scanner
     string.
 *)
 
+val bscanf_opt : Scanning.in_channel -> ('a, 'b, 'c, 'd) scanner_opt
+(** Same as {!Scanf.bscanf}, but returns [None] in case of scanning failure.
+
+    @since 5.00.0 *)
```
Comment (gasche) [inline stdlib/scanf.mli]: Answering this question is above my paygrade, but: is this the right version number?

**Comment by gasche** [comment]
I wonder if someone else would be willing to take this off Florian's shoulders here. Maybe one of @trefis, @Armael, @Engil?

File: stdlib/scanf.mli (hunk 1 above)
Comment (Octachron) [reply stdlib/scanf.mli]: The right version number should be `5.0`.

File: testsuite/tests/lib-scanf/tscanf.ml [hunk 2]
```diff
@@ -1555,3 +1555,12 @@ let test61 () =
...
 test (test61 ())
 ;;
+
+(* Tests for the option-returning variants *)
+let test62 () =
+  sscanf_opt "Hello" "%s" id = Some "Hello" &&
+  sscanf_opt "Hello" "%d" id = None &&
+  sscanf_opt "" "%d" id = None &&
+  sscanf_opt "Hello 123" "%s %d" (fun s n -> s, n) = Some ("Hello", 123) &&
+  sscanf_opt "Hello 123" "%s %r" (fun ib -> Scanf.bscanf_opt ib "%d" Fun.id) (fun s n -> s, n) = Some ("Hello", Some 123)
```
Comment (Octachron) [inline testsuite/tests/lib-scanf/tscanf.ml]: I would propose to also test that error in `%r` reader are also transformed into `None`:
```ocaml
sscanf_opt "Hello world" "%s %r" (fun ib -> Scanf.bscanf_opt ib "%d" Fun.id) (fun s n -> s, n) = None
```

**Review by Octachron — APPROVED** [review]
The code looks fine, and having a option version is definitively useful.

File: testsuite/tests/lib-scanf/tscanf.ml (hunk 2 above)
Comment (nojb) [reply testsuite/tests/lib-scanf/tscanf.ml]: Thanks, added.

File: stdlib/scanf.mli (hunk 1 above)
Comment (nojb) [reply stdlib/scanf.mli]: Thanks, fixed.
